### PR TITLE
[feature](nereids)support correlated scalar subquery without scalar agg

### PR DESCRIPTION
**Pull request by starocean999**
support correlated scalar subquery without scalar agg like:
`select t1.c1 from t1 
    where t1.c2 > (select t2.c2 from t2 where t1.c1 = t2.c1);`
after this pr, nereids would produce a correct plan for above sql.

### CODE REVIEW

**Comment by starocean999** [comment]
run buildall

**Comment by starocean999** [comment]
run buildall

**Comment by starocean999** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/SubqueryToApply.java [hunk 1]
```diff
@@ -383,32 +395,68 @@ private static boolean hasTopLevelScalarAgg(Plan plan) {
...
-                    .addAll(needAddScalarSubqueryOutputToProjects
-                        ? ImmutableList.of(subquery.getQueryPlan().getOutput().get(0)) : ImmutableList.of())
-                    .build();
-
-        return new LogicalProject(projects, newApply);
+        ImmutableList.Builder<NamedExpression> projects = ImmutableList.builder();
+        // left child
+        projects.addAll(childPlan.getOutput());
+        // markJoinSlotReference
+        projects.addAll(markJoinSlot.isPresent() ? ImmutableList.of(markJoinSlot.get()) : ImmutableList.of());
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/SubqueryToApply.java:444]:
```suggestion
        markJoinSlot.map(projects::add);
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/SubqueryToApply.java [hunk 2]
```diff
@@ -383,32 +395,68 @@ private static boolean hasTopLevelScalarAgg(Plan plan) {
...
-                    // markJoinSlotReference
-                    .addAll(subqueryToMarkJoinSlot.get(subquery).isPresent()
-                        ? ImmutableList.of(subqueryToMarkJoinSlot.get(subquery).get()) : ImmutableList.of())
-                    // scalarSubquery output
-                    .addAll(needAddScalarSubqueryOutputToProjects
-                        ? ImmutableList.of(subquery.getQueryPlan().getOutput().get(0)) : ImmutableList.of())
-                    .build();
-
-        return new LogicalProject(projects, newApply);
+        ImmutableList.Builder<NamedExpression> projects = ImmutableList.builder();
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/SubqueryToApply.java:440]:
```suggestion
        ImmutableList.Builder<NamedExpression> projects = ImmutableList.builderWithExpectedSize(childPlan.getOutput().size() + 3);
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/SubqueryToApply.java [hunk 3]
```diff
@@ -383,32 +395,68 @@ private static boolean hasTopLevelScalarAgg(Plan plan) {
...
+    private Pair<LogicalPlan, Optional<Expression>> addApply(SubqueryExpr subquery,
+            LogicalPlan childPlan,
+            Map<SubqueryExpr, Optional<MarkJoinSlotReference>> subqueryToMarkJoinSlot,
+            CascadesContext ctx, Optional<Expression> conjunct, boolean isProject,
+            boolean singleSubquery, boolean isMarkJoinSlotNotNull) {
         ctx.setSubqueryExprIsAnalyzed(subquery, true);
+        Optional<MarkJoinSlotReference> markJoinSlot = subqueryToMarkJoinSlot.get(subquery);
         boolean needAddScalarSubqueryOutputToProjects = isConjunctContainsScalarSubqueryOutput(
                 subquery, conjunct, isProject, singleSubquery);
+        boolean useNewSubquery = false;
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/SubqueryToApply.java:407]:
```suggestion
        boolean needRuntimeAssertCount = false;
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/SubqueryToApply.java [hunk 4]
```diff
@@ -383,32 +395,68 @@ private static boolean hasTopLevelScalarAgg(Plan plan) {
...
+                && !subquery.getCorrelateSlots().isEmpty()
+                && !((ScalarSubquery) subquery).hasTopLevelScalarAgg()) {
+            // if scalar subquery doesn't have top level scalar agg we will create one, for example
+            // select (select t2.c1 from t2 where t2.c2 = t1.c2) from t1;
+            // the original output of the correlate subquery is t2.c1, after adding a scalar agg, it will be
+            // select (select count(*), any_value(t2.c1) from t2 where t2.c2 = t1.c2) from t1;
+            Alias countAlias = new Alias(new Count());
+            Alias anyValueAlias = new Alias(new AnyValue(oldSubqueryOutput));
+            LogicalAggregate<Plan> aggregate = new LogicalAggregate<>(ImmutableList.of(),
+                    ImmutableList.of(countAlias, anyValueAlias), subquery.getQueryPlan());
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/SubqueryToApply.java:422]: group by will be add later?

**Comment by starocean999** [comment]
run buildall

**Comment by starocean999** [comment]
run buildall

**Comment by starocean999** [comment]
run buildall

**Comment by starocean999** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/SubExprAnalyzer.java [hunk 5]
```diff
@@ -114,26 +120,56 @@ public Expression visitInSubquery(InSubquery expr, T context) {
...
-
+        boolean isCorrelated = analyzedResult.isCorrelated();
+        LogicalPlan analyzedSubqueryPlan = analyzedResult.logicalPlan;
+        if (isCorrelated) {
+            if (analyzedSubqueryPlan instanceof LogicalLimit) {
+                if (ScalarSubquery.findTopLevelScalarAgg(analyzedResult.logicalPlan) == null) {
+                    throw new AnalysisException("limit is not supported in correlated subquery "
+                            + analyzedResult.getLogicalPlan());
+                } else {
+                    analyzedSubqueryPlan = (LogicalPlan) analyzedSubqueryPlan.child(0);
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/SubExprAnalyzer.java:131]: why just remove limit? what about limit offset? or limit 0?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/SubExprAnalyzer.java [hunk 6]
```diff
@@ -114,26 +120,56 @@ public Expression visitInSubquery(InSubquery expr, T context) {
...
                     && project.getProjects().size() == 1
                     && project.getProjects().get(0) instanceof Alias) {
+                // if scalar subquery is like select '2024-02-02 00:00:00'
+                // we can just return the constant expr '2024-02-02 00:00:00'
                 Alias alias = (Alias) project.getProjects().get(0);
                 if (alias.isConstant()) {
                     return alias.child();
                 }
+            } else if (isCorrelated) {
+                if (ExpressionUtils.containsWindowExpression(project.getProjects())) {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/SubExprAnalyzer.java:158]: add comment, it's not that easy to think about

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/SubExprAnalyzer.java [hunk 7]
```diff
@@ -245,15 +286,51 @@ public boolean hasGroupBy() {
...
+            ArrayDeque<Plan> planQueue = new ArrayDeque<>();
+            planQueue.add(rootPlan);
+            while (!planQueue.isEmpty()) {
+                Plan plan = planQueue.poll();
+                if (plan.getClass().equals(clazz)) {
+                    if (!Sets
+                            .intersection(slots,
+                                    ExpressionUtils.getInputSlotSet(plan.getExpressions()))
+                            .isEmpty()) {
+                        return true;
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/SubExprAnalyzer.java:316]: reformat it for easy reading for human

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/SubExprAnalyzer.java [hunk 8]
```diff
@@ -245,15 +286,51 @@ public boolean hasGroupBy() {
...
+                            ImmutableSet.copyOf(correlatedSlots), LogicalJoin.class);
+        }
+
+        public boolean hasCorrelatedSlotsInAgg() {
+            return correlatedSlots.isEmpty() ? false
+                    : findCorrelatedSlotsInNode(logicalPlan, ImmutableSet.copyOf(correlatedSlots),
+                            LogicalAggregate.class);
+        }
+
+        private static <T> boolean findCorrelatedSlotsInNode(Plan rootPlan,
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/SubExprAnalyzer.java:305]:
```suggestion
        private static <T> boolean hasCorrelatedSlotsInNode(Plan rootPlan,
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/SubExprAnalyzer.java [hunk 9]
```diff
@@ -172,6 +197,22 @@ private void checkNoCorrelatedSlotsUnderAgg(AnalyzedResult analyzedResult) {
...
+        }
+    }
+
+    private void checkNoCorrelatedSlotsInAgg(AnalyzedResult analyzedResult) {
+        if (analyzedResult.hasCorrelatedSlotsInAgg()) {
+            throw new AnalysisException(String.format(
+                    "outer query's column is not supported in subquery's aggregation operator : %s",
+                    analyzedResult.getCorrelatedSlots()));
+        }
+    }
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/SubExprAnalyzer.java:214]: add comment to explain the sql pattern or plan pattern want to check

File: regression-test/suites/nereids_p0/subquery/correlated_scalar_subquery.groovy [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+        insert into correlated_scalar_t1 values (1,null),(null,1),(1,2), (null,2),(1,3), (2,4), (2,5), (3,3), (3,4), (20,2), (22,3), (24,4),(null,null);
+    """
+    sql """
+        insert into correlated_scalar_t2 values (1,null),(null,1),(1,4), (1,2), (null,3), (2,4), (3,7), (3,9),(null,null),(5,1);
+    """
+    sql """
+        insert into correlated_scalar_t3 values (1,null),(null,1),(1,9), (1,8), (null,7), (2,6), (3,7), (3,9),(null,null),(5,1);
+    """
+
+    qt_select_where1 """select c1 from correlated_scalar_t1 where correlated_scalar_t1.c2 > (select c1 from correlated_scalar_t2 where correlated_scalar_t1.c1 = correlated_scalar_t2.c1 and correlated_scalar_t2.c2 < 4);"""
```
Comment (morrySnow) [inline regression-test/suites/nereids_p0/subquery/correlated_scalar_subquery.groovy:77]: add test case that sub query with
- limit
- limit offset
- order by
- topn
- topn offset
- corrlated filter + window + aggregate
- corrlated filter + lateral view + aggregate

**Comment by starocean999** [comment]
run buildall

**Comment by starocean999** [comment]
run buildall

**Comment by starocean999** [comment]
run buildall

**Comment by starocean999** [comment]
run buildall

**Comment by starocean999** [comment]
run buildall

**Review by morrySnow — DISMISSED** [review]

**Review by morrySnow — APPROVED** [review]

**Review by englefly — APPROVED** [review]

**Pull request merged**